### DISCUSSION TITLE
Async plugins and cross-requests

### DISCUSSION THREAD

**Discussion by miker059** [nuxt/nuxt#15896] [General]
Hello @danielroe and everybody!

Faced with the following problem. 

There is an SSR application, it uses Pinia for storage. 

Store initialization is implemented in server-side plugins. The authorization plugin works first. It checks for the presence of a refresh token in cookies and sends a request to receive an access token. Also, after receiving the access token, this plugin receives data about the user and fills in the user's store. 

The following plugin, using the user's data, his role, etc. initializes and fills in other stores necessary for the initial operation of the application, such as shopping cart, delivery addresses, etc. Recently, we conducted a small manual load testing by running the CURL script in several terminals, under different users, and noticed that for some requests, instead of 200 responses, 302 redirects were returned. We found out experimentally that it was a redirect inside auth middleware that triggered when the user is not logged in. After spending quite a lot of time searching, we found out that with asynchronous plugins, cross-requests behave strangely.

This is the answer we have built. It shows all the steps of initializing the application on the server side.

<img width="873" alt="Снимок экрана 2023-01-18 в 18 36 33" src="https://user-images.githubusercontent.com/6391978/213226479-8e2f72f0-b1dd-40cf-b487-84aa06f8a5b4.png">

For testing, we run a CURL script in two terminals that continuously sends requests to one url to our application from two different users. And we see that sometimes 302 arrives instead of 200 response code.

<img width="1392" alt="Снимок экрана 2023-01-18 в 18 23 01" src="https://user-images.githubusercontent.com/6391978/213230232-14f2409f-1189-4409-9aa1-10abf23d9e75.png">

Looking through the logs of the server side, we found the following:

<img width="1156" alt="Снимок экрана 2023-01-18 в 18 41 34" src="https://user-images.githubusercontent.com/6391978/213230826-fcacf3a4-d799-41b6-b714-6b156f526711.png">

In the screenshot above, we see that a request comes from a user with sessid 1674056224175, but immediately after comes a hung promise of the store initialization plugin from the previous user. After that, for some reason, the application decides that all plugins are initialized and passes control on. In the middleware, we see that the user's stop has not been initialized and because of this middleware redirects to the login page with the code 302.

Next we see the following:

<img width="1156" alt="Снимок экрана 2023-01-18 в 18 44 11" src="https://user-images.githubusercontent.com/6391978/213234296-94d8a9e9-b73d-4dca-9b52-9f55e2358356.png">

The session of another user starts and the completed plugin initialization promises of the previous user arrive in it.

An ordinary user, in principle, does not notice anything, since after redirecting to the login page, the server part sees the refresh token in cookies, makes a repeated request for access with a token and authorizes the user successfully. But we do not like this behavior, and we have already broken our whole head to understand what the reason is.

We also tried to put the initialization of plug-ins into the global middleware, but we ran into the same problem, asynchronous operations were ignored by the application and when a cross request from another user was sent to the next middleware, where there was a check for the fullness of the user's store, and since it was not filled, a redirect to the login with the code 302 occurred.

In App.vue, the placement is not suitable, since a completed list is required before the middleware.

I will be grateful for any help!

**Comment by danielroe** [comment]
This doesn't sound like an issue with Nuxt, as Nuxt very carefully scopes every composable to a single app instance that is unique per request. But I would be very happy to look into it further.

Would you provide a minimal reproduction, ideally not using pinia? (If you can't reproduce it without pinia, then it should probably be raised in the pinia repository.)

**Reply by miker059** [reply]
Indeed, the problem seems to be in Pinia. Today I changed the user's store to use useState, the redirect is gone, everything is working out as it should. Later today I will create a clean implementation with and without Pinia and demonstrate it here. 
@danielroe Thank You!
